Fixing segfault if filtered_cloud_topic is not set in PointCloudOctomapUpdater

jasedit opened the pull request:
Iterators were created into a PointCloud2 message which was only created if filtered_cloud_topic was set. Furthermore, the incoming data was never used during the update - the octomap update iterated over the output PointCloud2, as opposed to the input PointCloud 2. On my setup (Trusty/Indigo), PointCloudOctomapUpdater would crash immediately unless I defined filtered_cloud_topic, and I would never see any output data. I believe this pull request fixes this issue.

jasedit commented:
Odd - the build error is that perception/point_containment_filter/src/shape_mask.cpp could not find sensor_msgs/point_cloud2_iterator.h, which are not files I've touched in any way. It looks like CMakeLists.txt and packages.xml did not declare any dependency on sensor_msgs, which I've added. I think that should fix the Travis build error.

mikeferguson commented:
@jasedit Disregard build errors -- Travis doesn't support Trusty, so this is still being built against precise/hydro, and the iterator does not exist in hydro.

mikeferguson commented:
Looping in @vrabaud since he created the point_cloud2_iterator.

davetcoleman commented:
I've just run into this issue, I'll test your PR

Also, I've just disabled Travis for all MoveIt! repos per the mailing list email I sent out.

jasedit commented:
Thanks! I've verified that it no longer crashes on my system. However, I can't validate any of the data coming out of the updater due to #503, which I haven't made any headway on solving.

davetcoleman commented:
I'm still getting a segfault even with your branch when I start move_group with a point cloud being published:

> 0x00007fffd1abd503 in occupancy_map_monitor::PointCloudOctomapUpdater::cloudMsgCallback (this=0xa25f40, cloud_msg=...) at /home/dave/ros/ws_moveit/src/moveit_ros/perception/pointcloud_octomap_updater/src/pointcloud_octomap_updater.cpp:249

I'm currently debugging but any hints are welcomed

davetcoleman commented:
Also, it is not clear to me why you made the changes you made in this PR. Perhaps adding inline comments to your changes would help me and others in the future?

jasedit commented:
Line 249 would imply that cloud_msg is an empty/invalid message. There doesn't appear to be a check for this, and may be a reason why I haven't seen it yet.

jasedit commented:
I can add comments on the iterators momentarily.

jasedit commented:
Comments added - if there is additional info necessary, let me know and I'll be happy to flesh it out.

davetcoleman commented:
Thanks for the comments, I read a lot into the code just now. I believe your fixes are good but I think there are some more issues, in fact. See https://github.com/ros-planning/moveit_ros/issues/505.

In your testing are you actually filtering a point cloud, or do you have that feature disabled?

jasedit commented:
I am filtering a point cloud using this block of code. I don't think you can cause the segfault without filtering the points.

vrabaud commented on perception/pointcloud_octomap_updater/src/pointcloud_octomap_updater.cpp:
```diff
@@ -224,46 +233,45 @@ void PointCloudOctomapUpdater::cloudMsgCallback(const sensor_msgs::PointCloud2::
-        iter_y += point_subsample_, iter_z += point_subsample_)
+      sensor_msgs::PointCloud2ConstIterator<float> pt_iter(*cloud_msg, "x");
+      //set iterator to point at start of the current row
+      pt_iter += row;
```
that is wrong, you should have 

```
t_iter += row*cloud_msg->width;
```

jasedit replied:
Or `pt_iter += row_c`, no?

vrabaud replied:
that works too.

jasedit commented:
Commit is fixed to correctly address #505, I believe.

mikeferguson commented:
Ok, here's my thoughts:
- +1 to all changes in pointcloud_octomap_updater.cpp
- +1 to adding depend on sensor_msgs
- -0 to adding depends in cmake on sensor_msgs_generate_messages_cpp, there are a lot of other things that aren't declared, and most people aren't building sensor_msgs locally

I would recommend squashing this down to a single commit also.

jasedit commented:
Commits have been squashed, sensor_msgs_generate_messages_cpp has been removed.

davetcoleman commented on perception/point_containment_filter/CMakeLists.txt:
```diff
@@ -4,6 +4,7 @@ add_library(${MOVEIT_LIB_NAME} src/shape_mask.cpp)
 set_target_properties(${MOVEIT_LIB_NAME} PROPERTIES COMPILE_FLAGS "${CMAKE_CXX_FLAGS} ${OpenMP_CXX_FLAGS}")
 set_target_properties(${MOVEIT_LIB_NAME} PROPERTIES LINK_FLAGS "${OpenMP_CXX_FLAGS}")
 target_link_libraries(${MOVEIT_LIB_NAME} ${catkin_LIBRARIES} ${Boost_LIBRARIES})
+add_dependencies(${MOVEIT_LIB_NAME} sensor_msgs_generate_messages_cpp)
```
You said you took this out, right?

davetcoleman commented on perception/pointcloud_octomap_updater/src/pointcloud_octomap_updater.cpp:
```diff
@@ -224,46 +233,45 @@ void PointCloudOctomapUpdater::cloudMsgCallback(const sensor_msgs::PointCloud2::
     /* do ray tracing to find which cells this point cloud indicates should be free, and which it indicates
      * should be occupied */
-    for (unsigned int row = 0; row < cloud_msg->height; row += point_subsample_)
+    for(unsigned int row = 0; row < cloud_msg->height; row += point_subsample_)
```
This is super minor, but a digression from this file's standard

davetcoleman commented:
It works - thanks so much! I'll merge as soon as my two comments are addressed.

jasedit committed:
Fixing invalid iterators if filtered_cloud_topic is not set.
Adding missing dependency on sensor_msgs.
Fixing indentation, whitespace, and tabs.
Incrementing PointCloud2Iterator pixel-at-a-time, not byte-at-a-time.

jasedit commented:
Addressed!

jasedit commented:
Thanks for helping get to the bottom of this!

On Thu, Oct 23, 2014 at 2:10 AM, Dave Coleman notifications@github.com
wrote:

> Merged #504 https://github.com/ros-planning/moveit_ros/pull/504.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/ros-planning/moveit_ros/pull/504#event-182545205.
